xlsx/src/import/shared_strings.rs

```diff
@@ -1,39 +1,121 @@
+            current_string: String::new(),
+            strings: vec![],
+        }
+    }
+
+    fn process(&mut self, ev: Event) -> Result<(), XlsxError> {
+        self.state = match self.state {
+            ParserState::OutsideSi => match ev {
+                Event::Start(e) if e.local_name().into_inner() == b"si" => ParserState::InsideSi,
+                _ => ParserState::OutsideSi,
+            },
+            ParserState::InsideSi => match ev {
+                Event::Start(e) if e.local_name().into_inner() == b"t" => ParserState::T,
+                Event::End(e) if e.local_name().into_inner() == b"si" => {
+                    self.strings.push(self.current_string.clone());
+                    self.current_string.clear();
+                    ParserState::OutsideSi
+                }
+                _ => ParserState::InsideSi,
+            },
+            ParserState::T => match ev {
+                Event::Text(t) => {
+                    self.current_string
+                        .push_str(t.unescape().unwrap_or("".into()).as_ref());
+                    ParserState::T
+                }
+                Event::End(e) if e.local_name().into_inner() == b"t" => ParserState::InsideSi,
+                _ => ParserState::T,
+            },
+        };
+
+        Ok(())
+    }
+
+    fn strings(self) -> Result<Vec<String>, XlsxError> {
+        if self.state != ParserState::OutsideSi {
+            return Err(XlsxError::Xml("Corrupt XML structure".to_string()));
+        }
+        Ok(self.strings)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum ParserState {
+    OutsideSi,
+    InsideSi,
+    T,
+}
+
 /// Reads the list of shared strings in an Excel workbook
 /// Note than in IronCalc we lose _internal_ styling of a string
 /// See Section 18.4
 pub(crate) fn read_shared_strings<R: Read + std::io::Seek>(
     archive: &mut zip::read::ZipArchive<R>,
 ) -> Result<Vec<String>, XlsxError> {
     match archive.by_name("xl/sharedStrings.xml") {
-        Ok(mut file) => {
-            let mut text = String::new();
-            file.read_to_string(&mut text)?;
-            read_shared_strings_from_string(&text)
-        }
+        Ok(mut file) => read_shared_strings_from_reader(&mut file),
         Err(_e) => Ok(Vec::new()),
     }
 }
 
-fn read_shared_strings_from_string(text: &str) -> Result<Vec<String>, XlsxError> {
-    let doc = roxmltree::Document::parse(text)?;
-    let mut shared_strings = Vec::new();
-    let nodes: Vec<Node> = doc.descendants().filter(|n| n.has_tag_name("si")).collect();
-    for node in nodes {
-        let text = node
-            .descendants()
-            .filter(|n| n.has_tag_name("t"))
-            .map(|n| n.text().unwrap_or("").to_string())
-            .collect::<Vec<String>>()
-            .join("");
-        shared_strings.push(text);
+fn read_shared_strings_from_reader<R: Read>(reader: &mut R) -> Result<Vec<String>, XlsxError> {
+    let streaming = true;
+    if streaming {
```

Comment:
Right now this is how I'm keeping both implementations around. Should we do this in a different way, or remove the old implementation? I left them both here for now in case we want to do further perf / correctness testing. Changing `streaming` to false gives the old tree-based parsing behavior for shared strings.

Reply:
Removed it in latest commit.